### PR TITLE
Add extractinv pass, use it for synth_xilinx -ise

### PR DESCRIPTION
**Pull request by koriakin**

### CODE REVIEW

File: techlibs/xilinx/synth_xilinx.cc [hunk 1]
```diff
@@ -461,9 +461,10 @@ struct SynthXilinxPass : public ScriptPass
 				else
 					run("clkbufmap -buf BUFG O:I");
 			}
-			if (do_iopad)
+			if (help_mode || do_iopad)
 				run("iopadmap -bits -outpad OBUF I:O -inpad IBUF O:I A:top", "(only if '-iopad' or '-ise' and not '-noiopad')");
-		}
```
[the remark this answers is not in the document]
Comment (koriakin) [reply techlibs/xilinx/synth_xilinx.cc]: Whoops, conflict resolution mistake.

**Comment by koriakin** [comment]
Everything here could in principle be done by techmap, true.  However, this would require huge map files produced via some generator (because there are so many invertible pins in xilinx libraries).

I think we should ignore the _1 cells: the whole point of this pr is to emit something that ise accepts, and inv is already fine for that. The sketchy coverage of _1's is imo proof that we should consider them deprecated.

As for the polarity-style parameters: I think that's an overgeneralization.  This pass is meant to work with xilinx-style inverted attributes, and we don't have much choice about this.  Also, note that "active polarity" doesn't really apply to many signals covered here (eg. the multi-bit dsp opmode).

**Comment by koriakin** [comment]
@eddiehung it seems you replied to my post by editting it; please don't do that, for one there are no notifications for this.

> > Everything here could in principle be done by techmap, true.  However, this would require huge map files produced via some generator (because there are so many invertible pins in xilinx libraries).
> 
> Personally, I would lean towards a techmap file, even if it is a little less elegant. Once it is generated, anyone can easily customise the mapping rules in a more flexible manner (which leads to my other example...)
> I'm also a bit of a purist in that I would prefer not to introduce and maintain code that can be achieved using tools that are already present.

Fine, I'll make an alternate PR, but it's not going to be pretty.  Also, if we ever want to go the other direction (merge inverters back into cell parameters, perhaps as a preprocessing step for our own P&R), techmap won't help and we'll end up with a custom pass anyway.

> > I think we should ignore the _1 cells: the whole point of this pr is to emit something that ise accepts, and inv is already fine for that. The sketchy coverage of _1's is imo proof that we should consider them deprecated.
> 
> Yes, I would consider them deprecated. However, Yosys already emits `FD*_1` in the current 7 series flow (rather than using `IS_C_INVERTED`:
> https://github.com/YosysHQ/yosys/blob/eef0676105ff592e0e96bb835f0139f2f40d55bb/techlibs/xilinx/ff_map.v#L25
> 
> 
> possibly because of restrictions in downstream tools like VPR.
> I'm not saying this is something we want to keep up forever, either, but an choice should be provided for those who have to live with these restrictions.

So what? Any downstream consumer that does not do IS_*_INVERTED has to cope with INV cells anyway, because not all cells with invertible inputs have _1 variants.  Emitting _1 for some cells and INV for others is pointless.

> > As for the polarity-style parameters: I think that's an overgeneralization.  This pass is meant to work with xilinx-style inverted attributes, and we don't have much choice about this.  Also, note that "active polarity" doesn't really apply to many signals covered here (eg. the multi-bit dsp opmode).
> 
> True. But if you're building a new pass it would be nice to make it general purpose. One may need to operate on the Yosys internal flop models because they may want to extract inverters before flops are mapped to arch specific primitives so that those inverters can get rolled into techmapping.

Alright, how would you want to to work?  Alternate attribute name for polarity-style parameters? What if the polarity parameter is missing, should it default to all-1?  Would you still want the pass to delete the parameter once it's done, like it does now for inversion parameters (so that they don't reach ISE)?

> Which brings me to the related thought: looking at your modification to `synth_xilinx`, it appears that you call `extractinv` at the very end of the flow, introducing new `INV` cells. For non clock pins, wouldn't this mean this gets implemented as a `LUT1`?

No, non-clock pins with IS_*_INVERTED attributes have hard inverters as well — the INV cell is ephemeral and will be merged into the primitive (corresponding to <pin>INV attribute in XDL).

The `extractinv` is at the very end of the flow, because we want to do it after `clkbufmap` — otherwise, it will attempt to insert BUFGs between the INV and the cell, which we do not want.

**Comment by eddiehung** [comment]
> @eddiehung it seems you replied to my post by editting it; please don't do that, for one there are no notifications for this.

Really sorry about that. I had no idea I did that... I think I am losing it. I think I meant to click quote instead of edit.

> Also, if we ever want to go the other direction (merge inverters back into cell parameters, perhaps as a preprocessing step for our own P&R), techmap won't help and we'll end up with a custom pass anyway.

I had thought about that, and the intention was to develop a pattern matcher pass to fold inverters inwards.

**Comment by koriakin** [comment]
Simplified the code a bit as discussed in zoom, allowed non-blackboxes as well, resolved conflicts.

**Pull request merged**